core/types/calls.cc

```diff
@@ -2310,6 +2310,126 @@ class Tuple_concat : public IntrinsicMethod {
     }
 } Tuple_concat;
 
+namespace {
+
+optional<size_t> indexForKey(const ShapeType &shape, const LiteralType &argLit) {
+    auto fnd = absl::c_find_if(
+        shape.keys, [&argLit](auto &elemLit) { return argLit.equals(cast_type_nonnull<LiteralType>(elemLit)); });
+
+    if (fnd == shape.keys.end()) {
+        return nullopt;
+    } else {
+        return fnd - shape.keys.begin();
+    }
+}
+
+optional<Loc> locOfValueForKey(const GlobalState &gs, const Loc origin, const NameRef key, const TypePtr expectedType) {
+    if (!isa_type<ClassType>(expectedType)) {
+        return nullopt;
+    }
+    auto ct = cast_type_nonnull<ClassType>(expectedType);
+
+    // Unlike with normal `T.let` autocorrects, we don't have location information for "the value
+    // for a specific key". To make up for this, we hard-code the most common pinning errors by
+    // scannign the source directly.
+
+    const char *valueStr;
+    if (ct.symbol == core::Symbols::NilClass()) {
+        valueStr = "nil";
+    } else if (ct.symbol == core::Symbols::TrueClass()) {
+        valueStr = "true";
+    } else if (ct.symbol == core::Symbols::FalseClass()) {
+        valueStr = "false";
+    } else {
+        return nullopt;
+    }
```

Comment:
These three cases were enough to fix ~80% of the existing problems on Stripe's codebase.